backend/src/v2/driver/driver.go

```diff
@@ -2006,7 +2028,253 @@ func resolveUpstreamArtifacts(cfg resolveUpstreamOutputsConfig) (*pipelinespec.A
 	}
 }
 
-// provisionOuutputs prepares output references that will get saved to MLMD.
+func isCollection(cfg resolveUpstreamOutputsConfig, currentTask *metadata.Execution) ([]string, error) {
+	starterTaskQueue := make([]string, 0)
+	parentDAG, err := cfg.mlmd.GetExecution(cfg.ctx, currentTask.GetExecution().GetCustomProperties()["parent_dag_id"].GetIntValue())
+	if err != nil {
+		return nil, err
+	}
+	resolveCollection, iterations := getParallelForIterationCount(currentTask, parentDAG)
+
+	glog.V(4).Infof("resolveCollection: %v", resolveCollection)
+	glog.V(4).Infof("iterations: %v", iterations)
+	if resolveCollection {
+		for i := range iterations {
+			starterTaskName := metadata.UpdateTaskNameWithDagID(currentTask.TaskName(), currentTask.GetID())
+			starterTaskName = metadata.GetParallelForTaskName(starterTaskName, i)
+			starterTaskQueue = append(starterTaskQueue, starterTaskName)
+		}
+	}
+	return starterTaskQueue, nil
+}
+
+// In order to properly perform artifact/parameter collection across a DAG/tree
+// we perform a breadth first search traversal.
+func CollectInputs(cfg resolveUpstreamOutputsConfig, starterTaskQueue []string, tasks map[string]*metadata.Execution, outputKey string, isArtifact bool) (*structpb.Value, *pipelinespec.ArtifactList, error) {
+	glog.V(4).Infof("currentTask is a ParallelFor DAG. Attempting to gather all nested producer_subtasks")
+	// Set some helpers for the BFS
+	parallelForParameterList := make([]*structpb.Value, 0)
+	parallelForArtifactList := make([]*pipelinespec.RuntimeArtifact, 0)
+	resolveTaskNameQueue := starterTaskQueue
+	var currentTask *metadata.Execution
+	var workingSubTaskName string
+	workingOutputKey := outputKey
+	tempWorkingOutputKey := outputKey
+	previousTaskName := tasks[resolveTaskNameQueue[0]].TaskName()
```

Comment:
This should likely check the length of `resolveTaskNameQueue` before accessing its index.